lib/connections/limiter.go

```diff
@@ -162,3 +263,37 @@ func (b *atomicBool) set(v bool) {
 func (b *atomicBool) get() bool {
 	return atomic.LoadInt32((*int32)(b)) != 0
 }
+
+// Utility functions for atomic operations on device limiters map
+func (lim *limiter) getOrCreateDeviceWriteLimiter(deviceID protocol.DeviceID) (*rate.Limiter, bool) {
+	lim.deviceMapMutex.Lock()
+	defer lim.deviceMapMutex.Unlock()
+	limiter, ok := lim.deviceWriteLimiters[deviceID]
+
+	if !ok {
+		limiter = rate.NewLimiter(rate.Inf, limiterBurstSize)
+		lim.deviceWriteLimiters[deviceID] = limiter
+	}
+
+	return limiter, ok
+}
+
+func (lim *limiter) getOrCreateDeviceReadLimiter(deviceID protocol.DeviceID) (*rate.Limiter, bool) {
+	lim.deviceMapMutex.Lock()
+	defer lim.deviceMapMutex.Unlock()
+	limiter, ok := lim.deviceReadLimiters[deviceID]
+
+	if !ok {
+		limiter = rate.NewLimiter(rate.Inf, limiterBurstSize)
+		lim.deviceReadLimiters[deviceID] = limiter
+	}
+
+	return limiter, ok
+}
+
+func (lim *limiter) deleteDeviceLimiters(deviceID protocol.DeviceID) {
```

Comment:
This is literally used in one place, just embed this code in that place and be done with it.